File: JS - Fundamentals/11. Regular Expressions - Exercise/furniture.js
function solve(array){
    // const pattern = /^>>(?<item>[A-Za-z]+)<<(?<price>[\d?.]+)!(?<quantity>\d+)/g;
    const pattern = /^(?<separator>[>]){2}(?<item>[A-Za-z]+)(?<SecSeparator>[<]{2})(?<price>[\d?.]+)(?<thirdSeparator>[!]){1}(?<quantity>\d+)/g;
    let currentLine = array.shift();
    let total = 0;
    const result = ['Bought furniture:'];

    while(currentLine !== 'Purchase'){
        while((match = pattern.exec(currentLine)) !== null){
            const item = match.groups.item;
            const price = Number(match.groups.price);
            const quantity = Number(match.groups.quantity);

            result.push(item);
            total += price*quantity;
        }

        currentLine = array.shift();
    }
    result.push(`Total money spend: ${total.toFixed(2)}`);

    return result.join('\n');  
}

console.log(solve([ '>>Sofa<<312.23!3', '>>TV<<300!5', '>Invalid<<!5', 'Purchase' ]));

File: Java Fundamentals/10. Regular Expressions - Exercises/src/Furniture.java
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class Furniture {
    public static void main(String[] args) {
        Scanner scan = new Scanner(System.in);

        String input = scan.nextLine();
        double totalMoney = 0.0;
        List<String> items = new ArrayList<>();

        String pattern = ">>(?<item>\\w+)<<(?<price>\\d+\\.?\\d+)!(?<quantity>\\d+)";
        Pattern regex = Pattern.compile(pattern);

        while(input.equals("Purchase") == false){
            Matcher match = regex.matcher(input);

            if(match.find()){
                items.add(match.group("item"));
                totalMoney += Double.parseDouble(match.group("price")) * Integer.parseInt(match.group("quantity"));
            }

            input = scan.nextLine();
        }

        System.out.println("Bought furniture:");

        if(items.size() > 0){
            System.out.println(String.join("\n", items));
        }

        System.out.printf("Total money spend: %.2f", totalMoney);
    }
}
